dcp-0008/dcp-0008.mediawiki

```diff
@@ -0,0 +1,342 @@
+directly involve transactions and scripts since they are both fundamental
+building blocks that virtually all ecosystem components work with in some way or
+another.
+
+===Hard Forks Preferred===
+
+The two primary methods for updating consensus rules are commonly referred to as
+soft forks and hard forks.  The key difference between the two methods is that
+soft forks involve changing the rules in a way that older nodes are not aware
+that the change has happened whereas hard forks require older nodes to upgrade
+in order to continue following the chain and therefore are fully aware when the
+rules have changed.
+
+This ability to trick older nodes into believing they are faithfully validating
+every transaction when in reality they are not completely undermines the primary
+reason for running a fully-validating node.  In some scenarios, it enables a
+certain class of attacks against people running those out of date nodes by
+taking advantage of the fact they are not actually validating things according
+to the latest consensus rules.
+
+The main takeaway is that hard forks are the preferred method and the Decred
+voting process allows stakeholders to seamlessly choose which side of a hard
+fork is the agreed upon winner in a transparent and cryptographically-provable
+way.
+
+===Existing Version Semantics===
+
+Primarily due to behavior inherited from less capable systems limited to soft
+forks, currently, all transaction and script versions newer than the
+actively-enforced versions are permitted by the consensus rules.
+
+In particular:
+
+* Transactions of newer unknown versions are treated the same as if they were the latest actively-enforced version
+* Newer unknown versions of scripts are treated as valid and therefore never execute
+
+Unfortunately, in addition to other issues associated with soft forks, this
+behavior also means ecosystem components are unable to rely on the version
+having any meaning in terms of exactly which rules should be applied since newer
+versions are permissible _before_ their semantics are defined by future
+consensus upgrades.  In other words, the existence of a transaction or script
+with a given version doesn't guarantee it actually adheres to rules that apply
+to that version since it might have been created prior to the definition of
+those rules.
+
+The changes specified by this proposal permit ecosystem components to safely and
+reliably associate a given version with a specific set of well-defined consensus
+rules.
+
+==Specification==
+
+All transactions and scripts with a version that is greater than their currently
+supported maximum values MUST be rejected until a future consensus rule change
+introduces a new version along with fully defining its semantics.
+
+As of the implementation of this specification, this means imposing the
+following two new rules:
+
+* All transactions with versions greater than 3 MUST be rejected
+* All scripts with versions greater than 0 MUST be rejected
```

Comment:
I think this should call out that it only applies to regular transactions and not stake transactions (based on the reference implementation).   That point is not discussed anywhere in the DCP other than the reference implementation section having a `!isStakeTx` guard, so it's probably also worth mentioning that the script version for stake transactions is enforced separately.

Reply:
Great observation!  While it is implicitly true since all of the stake transactions already enforce it, it is definitely not a new rule for stake transaction, so this should be explicit about that.  Updated.